packages/elements/src/internals/machines/sign-in.actors.ts

```diff
@@ -5,73 +5,151 @@ import type {
 import { assertIsDefined } from './utils/assert';
 
-export const createSignIn = fromPromise<SignInResource, WithClient<{ fields: SignInMachineContext['fields'] }>>(
-  ({ input: { client, fields } }) => {
-    const password = fields.get('password');
-    const identifier = fields.get('identifier');
-
-    const passwordParams = password?.value
-      ? {
-          password: password.value,
-          strategy: 'password',
-        }
-      : {};
-
-    return client.signIn.create({
-      identifier: identifier?.value as string,
-      ...passwordParams,
+// ================= createSignIn ================= //
+
+export type CreateSignInInput = WithClient<{ fields: SignInMachineContext['fields'] }>;
+
+export const createSignIn = fromPromise<SignInResource, CreateSignInInput>(({ input: { client, fields } }) => {
+  const password = fields.get('password');
+  const identifier = fields.get('identifier');
+
+  const passwordParams = password?.value
+    ? {
+        password: password.value,
+        strategy: 'password',
+      }
+    : {};
+
+  return client.signIn.create({
+    identifier: identifier?.value as string,
+    ...passwordParams,
+  });
+});
+
+// ================= authenticateWithRedirect ================= //
+
+export type AuthenticateWithRedirectInput = WithClerk<{
+  environment: EnvironmentResource | undefined;
+  strategy: AuthenticateWithRedirectParams['strategy'] | undefined;
+}>;
+
+export const authenticateWithRedirect = fromPromise<void, AuthenticateWithRedirectInput>(
+  async ({ input: { clerk, environment, strategy } }) => {
+    assertIsDefined(environment);
+    assertIsDefined(strategy);
+
+    return clerk.client.signIn.authenticateWithRedirect({
+      strategy,
+      redirectUrl: `${environment.displayConfig.signInUrl}/sso-callback`,
+      redirectUrlComplete: clerk.buildAfterSignInUrl(),
     });
   },
 );
 
-export const authenticateWithRedirect = fromPromise<
-  void,
-  WithClerk<{
-    environment: EnvironmentResource | undefined;
-    strategy: AuthenticateWithRedirectParams['strategy'] | undefined;
-  }>
->(async ({ input: { clerk, environment, strategy } }) => {
-  assertIsDefined(environment);
-  assertIsDefined(strategy);
-
-  return clerk.client.signIn.authenticateWithRedirect({
-    strategy,
-    redirectUrl: `${environment.displayConfig.signInUrl}/sso-callback`,
-    redirectUrlComplete: clerk.buildAfterSignInUrl(),
-  });
-});
+// ================= determineStartingFirstFactor ================= //
+
+export type DetermineStartingFirstFactorInput = {
+  supportedFactors: SignInFirstFactor[];
+  identifier: string | null;
+  preferredStrategy: PreferredSignInStrategy | undefined;
+};
 
-export const prepareFirstFactor = fromPromise<SignInResource, WithClient<WithParams<PrepareFirstFactorParams>>>(
-  ({ input }) => input.client.signIn.prepareFirstFactor(input.params),
+export const determineStartingFirstFactor = fromPromise<SignInFactor | null, DetermineStartingFirstFactorInput>(
```

Comment:
Does this need to be an actor? The helper is synchronous and doesn't throw. we could just as easily use it with `assign()` to set the context value.

Reply:
Updating in a follow-up PR